pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/ActivityAnswerCopierSql.java

```diff
@@ -0,0 +1,405 @@
+                break;
+            } case BOOLEAN: {
+                answerCopier = new BooleanAnswerCopier();
+                break;
+            } case NUMERIC: {
+                answerCopier = new NumericAnswerCopier();
+                break;
+            }
+            default: {
+                throw new DaoException(answerToCopy.getQuestionType() + " question type does not have copy support");
+            }
+        }
+
+        return (T)answerCopier.copyAnswer(getHandle(),
+                sourceQuestionStableId,
+                destinationQuestionStableId,
+                sourceInstanceId,
+                destinationInstanceId,
+                createdAtEpochMillis,
+                lastUpdatedAtEpochMillis, compositeCopyConfiguration);
+    }
+
+    class AnswerToCopy {
+
+        private long answerId;
+
+        private QuestionType questionType;
+
+        @JdbiConstructor
+        public AnswerToCopy(@ColumnName("answer_id") long answerId,
+                            @ColumnName("question_type_code") QuestionType questionType) {
+            this.answerId = answerId;
+            this.questionType = questionType;
+        }
+
+        public long getAnswerId() {
+            return answerId;
+        }
+
+        public QuestionType getQuestionType() {
+            return questionType;
+        }
+    }
+
+    /**
+     * An output from copying an answer, which includes
+     * the source answer id and the destination id at a minimum.
+     */
+    class AnswerCopyResult {
+
+        private final long sourceAnswerId;
+
+        private final long destinationAnswerId;
+
+        public AnswerCopyResult(long sourceAnswerId, long destinationAnswerId) {
+            this.sourceAnswerId = sourceAnswerId;
+            this.destinationAnswerId = destinationAnswerId;
+        }
+
+        public long getSourceAnswerId() {
+            return sourceAnswerId;
+        }
+
+        public long getDestinationAnswerId() {
+            return destinationAnswerId;
+        }
+    }
+
+    interface AnswerCopier<T extends AnswerCopyResult> {
+
+        /**
+         * Copies the answer to one question in an activity instance
+         * to a potentially different question in a different activity instance.
+         */
+        T copyAnswer(Handle handle,
+                     String sourceQuestionStableId,
+                    String destinationQuestionStableId,
+                    long sourceInstanceId,
+                                    long destinationInstanceId,
+                                    long createdAtEpochMillis,
+                                    long lastUpdatedAtEpochMillis,
+                     CompositeAnswerCopyConfiguration compositeCopyConfiguration);
+    }
+
+    /**
+     * Copies top-level answer and composite_answer_item (when directed by compositeCopyConfig)
+     */
+    abstract class BaseAnswerCopier<T extends AnswerCopyResult> implements AnswerCopier {
```

Comment:
Looks like for all subclasses `T` is `AnswerCopyResult`. Maybe this doesn't need to be generic?

Reply:
An earlier iteration had composite-specific data in a `CompositeCopyResult` which was interesting for a time.  I don't have a concrete scenario for when we might want to introduce a new subclass, just more of a vague "might be useful someday".  I'll de-genericize.